lib/varip.c

```diff
@@ -0,0 +1,76 @@
+#include <rte_ip.h>
+#include <rte_mbuf.h>
+
+#include "gatekeeper_varip.h"
+
+struct ipv6_opt_hdr {
+	uint8_t nexthdr;
+	uint8_t hdrlen;
+} __attribute__((packed));
+
+/* Fragmentation header. */
+struct frag_hdr {
+	uint8_t  nexthdr;
+	uint8_t  reserved;
+	uint16_t frag_off;
+	uint32_t identification;
+};
+
+#define ipv6_optlen(p) (((p)->hdrlen + 1) << 3)
+
+bool
+ipv6_ext_hdr(uint8_t nexthdr)
+{
+	/* Find out if nexthdr is an extension header or a protocol. */
+	return (nexthdr == NEXTHDR_HOP) ||
+		(nexthdr == NEXTHDR_ROUTING) ||
+		(nexthdr == NEXTHDR_FRAGMENT) ||
+		(nexthdr == NEXTHDR_AUTH) ||
+		(nexthdr == NEXTHDR_NONE) ||
+		(nexthdr == NEXTHDR_DEST);
+}
+
+int
+ipv6_skip_exthdr(const struct ipv6_hdr *ip6hdr, int remaining_len,
+	int start, uint8_t *nexthdrp, uint16_t *frag_offp)
+{
+	uint8_t nexthdr = *nexthdrp;
+
+	*frag_offp = 0;
+
+	while (ipv6_ext_hdr(nexthdr)) {
+		int hdrlen;
+		const struct ipv6_opt_hdr *hp =
+			(const struct ipv6_opt_hdr *)
+			((const uint8_t *)ip6hdr + start);
+
+		if (nexthdr == NEXTHDR_NONE)
+			return -1;
+
+		if (hp == NULL)
+			return -1;
```

Comment:
But even if the packet is truncated, I don't think `hp` will be `NULL`. It will be some reference to memory that is beyond the valid bytes of the packet.